### ISSUE TITLE
Mock requests

### ISSUE THREAD

**Issue by mgrybyk** [webdriverio/webdriverio#4079]
**Is your feature request related to a problem? Please describe.**
I'd like to be to mock requests similar to what [Cypress](https://docs.cypress.io/api/commands/route.html#Syntax) does.

**Describe the solution you'd like**
I'd like to have new service `wdio-request-proxy` (suggest your namings).

The service itself should work like a proxy: `browser -> wdio-request-proxy -> my-site`.

Dedicated proxy should be started for every running spec file. Once test is finished proxy should be killed and port marked as free.

config
```
services: [
    ["request-proxy", { startPort: 35000 }]
]
```

```
/**
*
* url: null // string or regexp or glob (minimatch) url to match. 
* responseData: null // response data, REQUIRED
* options - {                   
                     method: 'GET', // request's method, default GET
                     status: 200, // response statusCode, default 200
                     delay: 0, // delay before sending response, default 0
                     ... }
* responseCb - optional response body modifier
* /
function mock(url, responseData, options, responseCb) {
  // ...
  const response = responseCb(urlMatcher, responseData, originalRequest)
  // ...
}
```

Usage:
```
import WdioProxy from 'wdio-request-proxy'
// or it should be a part of browser object?

describe("main suite", () => {
    before(() => {
      // proxy starts on free port from available ports defined in config
      // idea is to have proxy per spec file (worker)
      // suggest your ideas if you know how to deal with parallel tests in a better way
      WdioProxy.start()

      // simple
      WdioProxy.mock("**/api/users", [{ id: 1, name: foo }, { id: 2, name: 'bar' }])

      // advanced
      WdioProxy.mock("**/api/users/*/update", { success: true }, { method: "POST"}, (urlMatcher, body) => {
        return {
                ...body,
                id: urlMatcher[1]
        }
      })
    })

    it("some test", () => {
      expect($$('ul.users li.user').length).eq(2)
    })
})
```

**Describe alternatives you've considered**
anything other solution is welcomed!

**Additional context**
there might be some issues with https traffic, might need to generate certificates first and add them to keystore.

**Comment by christian-bromann** [comment]
I think it might be a better idea to just create a service that adds commands to the browser scope so that you only need to do:

```js
browser.mock(...)
```

We can't use the proxy capability as we are not able to get HTTPS traffic this way. Therefor I would suggest to start with Chrome only and build this into the `@wdio/devtools-service` and use the [intercept capabilities](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setRequestInterception) of CDP.
An other option would be to use [mitmproxy](https://mitmproxy.org/) that allows to proxy even https requests and overwrite responses. However this will cause some dependency issues as we have to install the proxy.

IMO: I would love to see this in Chrome and hope that other browser will support request interception at some point too. Cypress will also only be able to support this in Chrome either.

**Comment by luiscspinho** [comment]
Hi,

We came across this problem when implementing our tests, how to intercept requests (HTTPS) in our application while using `wdio`, and I hope I can help you with our approach.
We use the `MockServer standalone`  dependency in order to intercept the requests (I guess that `mitmproxy` would work the same way), then in the `beforeSession` hook we spawn a node process of our `MockServer` on a random free port, which we will use to instantiate the `capabilities.proxy.[ssl/http proxy]` object and connect our browser to it. After this, each test will register intentions/mocks (calling `localhost:port`) in the proxy and that's it.

I think you get the idea, but I can create a small project working with this if you guys want. (**Very, very** personal take, but I would prefer to have documentation about how to do something like this than to have a @wdio/service doing it for me, creating dependencies between wdio and browsers/proxys etc.)

**Comment by kevinmcdonnell** [comment]
@luiscspinho could you share an example of how to do this? I'd like to try this and I'd love to see a working example

**Comment by luiscspinho** [comment]
Hi @kevinmcdonnell ,
I've created this small example: https://github.com/luiscspinho/webdriverio-mock/
The only thing missing is downloading the [MockServer](http://www.mock-server.com/where/downloads.html) `jar` and putting it on the project root so the `mockserver` npm task can find it. Any doubts/improvements are welcome 👍

**Comment by nanomosfet** [comment]
@luiscspinho the Java dependency that it adds could be a problem for some. At that point you can just use the [wiremock service](https://webdriver.io/docs/wdio-wiremock-service.html)

**Comment by luiscspinho** [comment]
@nanomosfet Well I guess it's the same thing, the wiremock service also adds a JAR dependency ( https://github.com/erwinheitzman/wdio-wiremock-service/blob/master/launcher.js#L40 ) but you don't see it, the same could be done to the mockserver. I totally get your point, that's why I said that I'm not sure this is the way to go before. However I'm using this solution on my tests and loving it so far

**Comment by nanomosfet** [comment]
yeah. I would like to see a built in version from the browser object like @christian-bromann said

> I think it might be a better idea to just create a service that adds commands to the browser scope so that you only need to do:
> 
> browser.mock(...)
> We can't use the proxy capability as we are not able to get HTTPS traffic this way. Therefor I would suggest to start with Chrome only and build this into the @wdio/devtools-service and use the intercept capabilities of CDP.
> An other option would be to use mitmproxy that allows to proxy even https requests and overwrite responses. However this will cause some dependency issues as we have to install the proxy.
> 
> IMO: I would love to see this in Chrome and hope that other browser will support request interception at some point too. Cypress will also only be able to support this in Chrome either.

even if its only for chrome i think it would be worth it because `cypress` supports it out of the box. We should just warn people that it wont work for firefox

**Comment by christian-bromann** [comment]
> I would like to see a built in version from the browser object like @christian-bromann said

@nanomosfet you can already use the devtools package to mock the API using Puppeteer. Providing a better mock and stubbing functionality is next on my list 😉

**Comment by christian-bromann** [comment]
closing in favor of #5336
